### PR TITLE
Autofix Markdown files using Prettier on merge

### LINKED ISSUE #3256: Prettier should automatically fix and commit markdown files
**Status:** Open for comments

## Need

This might be controversial so I added the RFC tag 😄 

The Backstage docs have this "Edit" button, which is (I guess) supposed to make it easy to update the docs using the GitHub UI. I ❤️  this idea and I like to use it. Unfortunately, 2/3 times so far that I have tried to submit changes to docs, prettier fails for me and I have to go and clone the branch locally, run prettier and force push.

## Proposal

 1. Prettier should automatically run in a GitHub action against markdown files in the `/docs` directory.
 2. The `--write` flag should be enabled enabled.
 3. The result should be automatically committed to fix formatting.

## Alternatives

 1. I could try harder to do what prettier demands of me.

## Risks

Automated commits always come with some risk. I feel like the risk can be controlled by limiting the filetype and directories that this runs in.

### PR DESCRIPTION
**Pull request by tudi2d**
Add seperate workflow for autofixing MD files using Prettier when PRs get merged into master. Use case: Edit docs from browser, but they don't align Prettier config. For more info checkout out the linked issue

Closes #3256

#### :heavy_check_mark: Checklist

- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by tudi2d** [comment]
Would be happy about feedback. Not 100% sure how to fully test this, as "CI / verify" will run `yarn prettier:check` & this would fail for an not prettified "ADOPTERS.md".

**Review by TejasQ — APPROVED** [review]
Looks great to me!

File: .github/workflows/prettify.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            ${{ runner.os }}-yarn-
+      - name: yarn install
+        run: yarn install --frozen-lockfile
+      # End of yarn setup
+
+      - name: Run Prettier on docs
+        uses: creyD/prettier_action@v3.1
+        with:
+          # Modifies commit only if prettier autofixed a Markdown file
+          prettier_options: --config docs/prettier.config.js --write docs/**/*.md
```
Comment (Rugvip) [inline .github/workflows/prettify.yml:50]: Let's try only running this towards `ADOPTERS.md` for now, to try out the flow

**Review by Rugvip — COMMENTED** [review]
Yeah I think the autofix bit looks good, although as mentioned let's just do it for `ADOPTERS.md` for now.

The missing bit is to ignore prettier errors in the files that will be auto formatted in the CI check, because otherwise PRs will fail and never allow this auto fix to be run.

The best way to test actions is really to create a separate repository where you try them. In this case I don't think that'll be needed tbh, just wanna be somewhat confident and then merge and try it out here.

File: .github/workflows/prettify.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+on:
+  push:
+    branches:
+      - master
+
+jobs:
+  autofix-markdown:
+    name: Autofix Markdown files using Prettier
+    runs-ons: ubuntu-latest
```
Comment (freben) [inline .github/workflows/prettify.yml:11]: this is a typo, right?

File: .github/workflows/prettify.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        with:
+          path: ${{ steps.yarn-cache.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+      - name: yarn install
+        run: yarn install --frozen-lockfile
+      # End of yarn setup
+
+      - name: Run Prettier on docs
```
Comment (freben) [inline .github/workflows/prettify.yml:46]: this isn't on docs though

File: .github/workflows/prettify.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          only_changed: true
+          commit_message: 'Autofix documentation Markdown files using Prettier'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Run Prettier on root
+        uses: creyD/prettier_action@v3.1
+        with:
+          # Modifies commit only if prettier autofixed a Markdown file
+          prettier_options: --write *.md
```
Comment (freben) [inline .github/workflows/prettify.yml:60]: hm i'm not sure about this one - does it run non-recursively, on all files in the root, including adopters? But without the config file? I'm not entirely sure about this block

Comment (tudi2d) [reply .github/workflows/prettify.yml]: Good catch - this was a duplicate block - I removed it.

**Review by freben — APPROVED** [review]

**Pull request merged**